datashader/datatypes.py

```diff
@@ -0,0 +1,364 @@
+                    ['mask', 'start_indices', 'flat_array'])):
+
+            self._mask = data['mask']
+            self._start_indices = data['start_indices']
+            self._flat_array = data['flat_array']
+        else:
+            # Compute lengths
+            index_len = len(data)
+            buffer_len = sum(len(datum)
+                             if datum is not None
+                             else 0 for datum in data)
+
+            # Compute necessary precision of start_indices array
+            for nbits in [8, 16, 32, 64]:
+                start_indices_dtype = 'uint' + str(nbits)
+                max_supported = np.iinfo(start_indices_dtype).max
+                if buffer_len <= max_supported:
+                    break
+
+            # infer dtype if not provided
+            if dtype is None:
+                dtype = np.result_type(*[np.atleast_1d(v)
+                                         for v in data
+                                         if v is not None])
+
+            # Initialize representation arrays
+            self._mask = np.zeros(index_len, dtype='bool')
+            self._start_indices = np.zeros(index_len, dtype=start_indices_dtype)
+            self._flat_array = np.zeros(buffer_len, dtype=dtype)
+
+            # Populate arrays
+            next_start_ind = 0
+            for i, array_el in enumerate(data):
+                # Check for null values
+                isnull = array_el is None
+
+                # Compute element length
+                n = len(array_el) if not isnull else 0
+
+                # Update mask
+                self._mask[i] = isnull
+
+                # Update start indices
+                self._start_indices[i] = next_start_ind
+
+                # Update flat array
+                self._flat_array[next_start_ind:next_start_ind+n] = array_el
+
+                # increment next start index
+                next_start_ind += n
+
+    @property
+    def flat_array(self):
+        """
+        numpy array containing concatenation of all nested arrays
+
+        Returns
+        -------
+        np.ndarray
+        """
+        return self._flat_array
+
+    @property
+    def mask(self):
+        """
+        boolean numpy array the same length as the ragged array where values
+        of True indicate missing values.
+
+        Returns
+        -------
+        np.ndarray
+        """
+        return self._mask
+
+    @property
+    def start_indices(self):
+        """
+        integer numpy array the same length as the ragged array where values
+        represent the index into flat_array where the corresponding ragged
+        array element begins.
+
+        Returns
+        -------
+        np.ndarray
+        """
+        return self._start_indices
+
+    def __len__(self):
+        """
+        Length of this array
+
+        Returns
+        -------
+        length : int
+        """
+        return len(self._start_indices)
+
+    def __getitem__(self, item):
```

Comment:
What do you think should be the return value here (using the RaggedArray from your original post?

```python
In [25]: ra[[0]]
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-25-9a9c4b76d0ab> in <module>
----> 1 ra[[0]]

~/sandbox/datashader/datashader/datatypes.py in __getitem__(self, item)
    184             return RaggedArray(data, dtype=self.flat_array.dtype)
    185         else:
--> 186             raise KeyError(item)
    187
    188     @classmethod

KeyError: [0]
```

I believen that should be (possibly)

```python
In [29]: np.atleast_2d(ra[0])
Out[29]: array([[1., 2.]])
```

though still thinking things through.

Reply:
Yeah, I agree that is should be `array([[1., 2.]])`, I'll take a look at what's going wrong.

Reply:
Ah, yes it should indeed be a length-1 RaggedArray.